derivation/shared/src/test/scala/io/circe/derivation/JsonCodecMacrosSuite.scala

```diff
@@ -0,0 +1,159 @@
+package io.circe.derivation
+
+import cats.instances.AllInstances
+import cats.kernel.Eq
+import io.circe.{Decoder, Encoder, ObjectEncoder}
+import io.circe.derivation.jsoncodecmacrossuiteaux._
+import io.circe.testing.{ ArbitraryInstances, CodecTests }
+import org.scalacheck.{ Arbitrary, Gen }
+
+package object jsoncodecmacrossuiteaux extends AnyRef
+  with AllInstances with ArbitraryInstances
+
+package jsoncodecmacrossuiteaux {
+
+  // Simple
+
+  @JsonCodec final case class Simple(i: Int, l: Long, s: String)
+
+  object Simple {
+    implicit def eqSimple: Eq[Simple] = Eq.by(s => (s.i, s.l, s.s))
+
+    implicit def arbitrarySimple: Arbitrary[Simple] =
+      Arbitrary(
+        for {
+          i <- Arbitrary.arbitrary[Int]
+          l <- Arbitrary.arbitrary[Long]
+          s <- Arbitrary.arbitrary[String]
+        } yield Simple(i, l, s)
+      )
+  }
+
+  // Single
+
+  @JsonCodec final case class Single(i: Int)
+
+  object Single {
+    implicit def eqSingle: Eq[Single] = Eq.by(_.i)
+
+    implicit def arbitrarySingle: Arbitrary[Single] =
+      Arbitrary(Arbitrary.arbitrary[Int].map(Single(_)))
+  }
+
+  // Typed1
+
+  @JsonCodec final case class Typed1[A](i: Int, a: A, j: Int)
+
+  object Typed1 {
+    implicit def eqTyped1[A: Eq]: Eq[Typed1[A]] = Eq.by(t => (t.i, t.a, t.j))
+
+    implicit def arbitraryTyped1[A](implicit A: Arbitrary[A]): Arbitrary[Typed1[A]] =
+      Arbitrary(
+        for {
+          i <- Arbitrary.arbitrary[Int]
+          a <- A.arbitrary
+          j <- Arbitrary.arbitrary[Int]
+        } yield Typed1(i, a, j)
+      )
+  }
+
+  // Typed2
+
+  @JsonCodec final case class Typed2[A, B](i: Int, a: A, b: B, j: Int)
+
+  object Typed2 {
+    implicit def eqTyped2[A: Eq, B: Eq]: Eq[Typed2[A, B]] = Eq.by(t => (t.i, t.a, t.b, t.j))
+
+    implicit def arbitraryTyped2[A, B](implicit A: Arbitrary[A], B: Arbitrary[B])
+    : Arbitrary[Typed2[A, B]] =
+      Arbitrary(
+        for {
+          i <- Arbitrary.arbitrary[Int]
+          a <- A.arbitrary
+          b <- B.arbitrary
+          j <- Arbitrary.arbitrary[Int]
+        } yield Typed2(i, a, b, j)
+      )
+  }
+
+  // Hierarchy
+
+//  @JsonCodec sealed trait Hierarchy
```

Comment:
Out of curiosity, why is this commented out?

Reply:
Oh, of course—reading too quickly this morning.

Reply:
No worries :) Out of curiosity, how do you plan on adding the support for ADTs?